classes/renderer/hasmany.php

```diff
@@ -153,77 +296,78 @@ protected static function getConfig($item, $data) {
+    /**
+     * Gets or generates the renderer unique id
+     *
+     * @return array|mixed|string
+     */
+    public function getId()
     {
-        $renderer_options = \Arr::merge(static::$DEFAULT_RENDERER_OPTIONS, $renderer_options);
-        static $auto_id_increment = 1;
-        $index = \Input::get('index', $index);
-        $config = static::getConfig($item, $data);
-        $fieldset = static::getFieldSet($config, $item);
-        // Override auto_id generation so it don't use the name (because we replace it below)
-        $auto_id = uniqid('auto_id_');
-        // Will build hidden fields seperately
-        $fields = array();
-        foreach ($fieldset->field() as $field) {
-            $field->set_attribute('id', $auto_id.$auto_id_increment++);
-            if ($field->type != 'hidden' || (mb_strpos(get_class($field), 'Renderer_') != false)) {
-                $fields[] = $field;
-            }
-        }
-
-
-        $fieldset->form()->set_config('field_template', '<tr><th>{label}</th><td>{field}</td></tr>');
-        $view_params = array(
-            'fieldset' => $fieldset,
-            'fields' => $fields,
-            'is_new' => $item->is_new(),
-            'index' => $index,
-            'options' => $renderer_options,
-        );
-        $view_params['view_params'] = &$view_params;
+        $id = $this->get_attribute('id');
+        return !empty($id) ? $id : uniqid('hasmany_');
+    }
 
-        $replaces = array();
-        foreach ($config['fieldset_fields'] as $name => $item_config) {
-            $replaces['"'.$name.'"'] = '"'.$relation.'['.$index.']['.$name.']"';
+    /**
+     * Returns the context of the specified $item
+     *
+     * @param $item
+     * @return bool
+     */
+    public function getItemContext($item)
+    {
+        if (empty($item)) {
+            return false;
         }
-        $return = (string) \View::forge('novius_renderers::hasmany/item', $view_params, false)->render();
-
-        \Event::trigger('novius_renderers.hasmany_fieldset');
-
-        \Event::trigger_function('novius_renderers.hasmany_fieldset', array(
-            array(
-                'item' => &$item,
-                'index' => &$index,
-                'relation' => &$relation,
-                'replaces' => &$replaces
-            )
-        ));
-
-        return strtr($return, $replaces);
+        if (!$item::behaviours('Nos\Orm_Behaviour_Contextable') && !$item::behaviours('Nos\Orm_Behaviour_Twinnable')) {
+            return false;
+        }
+        return $item->get_context();
     }
 
-    public static function js_init()
+    /**
+     * Returns the field name of the context for the specified $model
+     *
+     * @param $model
+     * @return bool
+     */
+    public function getItemContextField($model)
     {
-        return \View::forge('novius_renderers::hasmany/js', array(), false);
+        $field = \Arr::get($this->getItemContextProperties($model), 'context_property');
```

Comment:
It would have no sense, I added theses methods for the code to be more readable.